pisa/core/bin_indexing.py

```diff
@@ -0,0 +1,217 @@
+# Numba vectorized functions
+
+if FTYPE == np.float32:
+    _SIGNATURE = ['(f4[:], f4[:], f4[:], f4[:], f4[:])']
+else:
+    _SIGNATURE = ['(f8[:], f8[:], f8[:], f8[:], f8[:])']
+
+@guvectorize(_SIGNATURE, '(),(),(j),(k)->()', target=TARGET)
+def lookup_index_vectorized_2d(sample_x, sample_y, bin_edges_x, bin_edges_y, indices):
+    """Same as above, except we get back the index"""
+    sample_x_ = sample_x[0]
+    sample_y_ = sample_y[0]
+    if (sample_x_ >= bin_edges_x[0]
+            and sample_x_ <= bin_edges_x[-1]
+            and sample_y_ >= bin_edges_y[0]
+            and sample_y_ <= bin_edges_y[-1]):
+        idx_x = find_index(sample_x_, bin_edges_x)
+        idx_y = find_index(sample_y_, bin_edges_y)
+        idx = idx_x*(len(bin_edges_y)-1) + idx_y
+        indices[0] = idx
+    else:
+        indices[0] = 0.
+
+
+
+
+if FTYPE == np.float32:
+    _SIGNATURE = ['(f4[:], f4[:], f4[:], f4[:], f4[:], f4[:], f4[:])']
+else:
+    _SIGNATURE = ['(f8[:], f8[:], f8[:], f8[:], f8[:], f8[:], f8[:])']
+
+@guvectorize(_SIGNATURE, '(),(),(),(j),(k),(l)->()', target=TARGET)
+def lookup_index_vectorized_3d(sample_x, sample_y, sample_z,  bin_edges_x, bin_edges_y, bin_edges_z, indices):
+    """Vectorized gufunc to perform the lookup"""
+    sample_x_ = sample_x[0]
+    sample_y_ = sample_y[0]
+    sample_z_ = sample_z[0]
+    if (sample_x_ >= bin_edges_x[0]
+            and sample_x_ <= bin_edges_x[-1]
+            and sample_y_ >= bin_edges_y[0]
+            and sample_y_ <= bin_edges_y[-1]
+            and sample_z_ >= bin_edges_z[0]
+            and sample_z_ <= bin_edges_z[-1]):
+        idx_x = find_index(sample_x_, bin_edges_x)
+        idx_y = find_index(sample_y_, bin_edges_y)
+        idx_z = find_index(sample_z_, bin_edges_z)
+        idx = (idx_x*(len(bin_edges_y)-1) + idx_y)*(len(bin_edges_z)-1) + idx_z
+        indices[0] = idx
+    else:
+        indices[0] = 0.
+
+
+
+
+
+
+def test_lookup_indices():
+    """Unit tests for `histogram` function"""
+    n_evts = 100
+    x = np.array([1,1,1,1,1,3,4,5,6,7], dtype=FTYPE)
+    y = np.array([2,2,2,2,2,3,3,3,3,3], dtype=FTYPE)
+    z = np.array([0,0,0,0,1,0,0,0,0,0], dtype=FTYPE)
+
+    w = np.ones(n_evts, dtype=FTYPE)
+
+    x = SmartArray(x)
+    y = SmartArray(y)
+    z = SmartArray(z)
+
+    w = SmartArray(w)
+
+    binning_x = OneDimBinning(name='x', num_bins=7, is_lin=True, domain=[0, 7])
+    binning_y = OneDimBinning(name='y', num_bins=3, is_lin=True, domain=[0, 4])
+    binning_z = OneDimBinning(name='z', num_bins=2, is_lin=True, domain=[0, 1])
+    
+    binning_1d = MultiDimBinning([binning_x])
+    binning_2d = MultiDimBinning([binning_x, binning_y])
+    binning_3d = MultiDimBinning([binning_x, binning_y, binning_z])
+
+
+    # 1D case: check that each event falls into its predicted bin
+    print('TEST 1D:')
+    print('array in 1D: ',x.get(WHERE),'\nBinning: ',binning_1d.bin_edges)
+    indices = lookup_indices([x], binning_1d)
```

Comment:
can you here do an actual test checking wether the returned values are what's expected? E.g. using some sort of `assert indices == expected_indices`

Reply:
Hey!

Just added a couple of assert test matching the calculated binning with what one would expect from the binning logic (ie, I manually checked that each value in the initial arrays should fall into so and so bin, in the 1D, 2D and 3D case).

while doing so, Justin found a bug in `find_index`: (in translation.py)

The code had :
`if (x < bin_edges[i+1]) or (x <= bin_edges[-1] and i == len(bin_edges) - 1)`

but if `i== len(bin_edges)`, that  first test fails. Python evaluates the first condition and short-circuits if the conclusion is clear. If not, it tests subsequent conditions. So the following does work in regular python:

`if (x <= bin_edges[-1] and i == len(bin_edges) - 1) or x < bin_edges[i+1]`